Bio/UniProt/__init__.py

```diff
@@ -15,3 +15,106 @@
 See also Bio.SeqIO.SwissIO for the "uniprot-xml" support in
 Bio.SeqIO.
 """
+
+import json
+import re
+import urllib.parse
+from http.client import HTTPMessage
+
+from typing import Optional, Iterator, List
+from urllib.request import urlopen
+
+
+_re_next_link = re.compile(r'<(.+)>; rel="next"')
+
+
+def _get_next_link(headers: HTTPMessage) -> Optional[str]:
+    if "Link" in headers:
+        match = _re_next_link.match(headers["Link"])
+        if match:
+            return match.group(1)
+
+    return None
+
+
+class _UniProtSearchResultIterator:
+    """An iterator over the results of a UniProt search.
+
+    Do not use this class directly. Instead, use the :meth:`UniProt.search` method.
+    """
+
+    def __init__(self, first_url: str):
+        self.batch_iterator: Iterator[dict] = iter([])
+        self.next_url: Optional[str] = first_url
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> dict:
+        try:
+            return next(self.batch_iterator)
+        except StopIteration:
+            if self.next_url is not None:
+                with urlopen(self.next_url) as response:
+                    self.next_url = _get_next_link(response.headers)
+                    self.batch_iterator = iter(
+                        json.loads(response.read().decode())["results"]
+                    )
+                    return next(self.batch_iterator)
+            else:
+                raise StopIteration
```

Comment:
Does the response headers tell you how many results are returned? If so, we can add a `__len__` method to this iterator, returning the number of results without having to actually iterate over them.
Also, we may want to add a `total` or `total_count` attribute to show that total number of queries that would have been obtained if we had not limited `size` to 500.
If we want to be fancy, we could use the `cursor` argument to obtain the results on demand. Then we won't need a `count_search_results_for` function. In that case, `__len__` should return the total number of queries obtainable. Perhaps we want to do that later, and then I am a bit uneasy to include a `count_search_results_for` function now if we don't need it in the end (and will deprecate it).